Add new tail sampling processor policy: probabilistic

yvrhdn opened the pull request:
**Description:**
This adds a new tail sampling policy that samples a percentage of traces.

Commit:
Add new tail sampling processor policy: percentage

Commit:
Update CHANGELOG.md

Commit:
Fix tail_sampling_config.yaml

Commit:
Fix typo

Commit:
Reset counters to avoid overflow, improve tests

jpkrohling commented on processor/tailsamplingprocessor/sampling/percentage.go:
```diff
@@ -0,0 +1,81 @@
+	decision := NotSampled
+
+	if float32(r.tracesSampled)/float32(r.tracesProcessed) <= r.percentage {
+		r.tracesSampled++
```
You probably need a lock here: multiple routines might have access to the same policy instance.

jpkrohling commented on processor/tailsamplingprocessor/README.md:
```diff
@@ -15,6 +15,7 @@ Multiple policies exist today and it is straight forward to add more. These incl
 - `always_sample`: Sample all traces
 - `latency`: Sample based on the duration of the trace. The duration is determined by looking at the earliest start time and latest end time, without taking into consideration what happened in between.
 - `numeric_attribute`: Sample based on number attributes
+- `percentage`: Sample a percentage of traces. Only traces that have not been sampled yet by another policy are taken into account.
```
Why would I use this instead of the probabilistic sampling processor? 
https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor/probabilisticsamplerprocessor

kvrhdn replied:
Oh, wasn't aware of the probabilistic sampling processor (for some reason) 🤦🏻  I'll take a look at it and make sure our terminology is consistent.

I think the main difference is when the sampling decision happens. If you only want percentage or probabilistic sampling it doesn't make sense to use this tail sampling processor.
But combining this with other tail sampling policies makes it more meaningful. For instance:

```yaml
    policies:
      [
          {
            name: all-errors,
            type: status_code,
            status_code: {status_codes: [ERROR]}
          },
          {
            name: half-of-remaining,
            type: percentage,
            percentage: {percentage: 0.5}
          },
     ]
```

This pipeline would sample all traces with status code error and 50% of the remaining traces. Using the probabilistic sampling processor you risk dropping traces with errors.

jpkrohling replied:
Good point. Make sure to mention that in the readme then.

joe-elliott replied:
Is there a difference between this and just putting the probabilistic sampling processor after tail sampling at 50%?

kvrhdn replied:
The result will be slightly different: putting the probabilistic sampling processor after the tail sampling processor will filter what the tail sampling processor samples.
The percentage sampler as implemented here will filter what is not sampled by another policy.

So for instance a pipeline:
```
tail sampling (sample all errors) -> probabilistic (at 50%)
```

Result: 50% of traces with error
Why:
- the tail sampler drops every non-error
- the probabilistic sampler drops 50% of what the tail sampler returns

While the following:
```
tail sampling (sampler all errors -> sample 50%)
```

Result: all traces with error + 50% of traces without

kvrhdn commented:
Not stale, I've been busy with some other stuff but I will continue working on this soon 🤞🏻

Commit:
Merge branch 'main' into percentage_sampling

# Conflicts:
#	CHANGELOG.md

Commit:
Move into internal as well

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/percentage.go:
```diff
@@ -0,0 +1,81 @@
+	decision := NotSampled
+
+	if float32(r.tracesSampled)/float32(r.tracesProcessed) <= r.percentage {
+		r.tracesSampled++
```
Don't you need to shield this field to protect it from concurrent modifications?

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/percentage.go:
```diff
@@ -0,0 +1,81 @@
+		r.tracesSampled++
+		decision = Sampled
+	}
+	r.tracesProcessed++
```
Same here. You might want to use a compare-and-set operation instead.

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/percentage.go:
```diff
@@ -0,0 +1,81 @@
+	r.tracesProcessed++
+
+	// reset counters to avoid overflow
+	if r.tracesProcessed == 1000 {
```
Why 1000? If the lowest a user can set is 0.001, we should document this. Also, have you considered using a ring instead? You can use something like:

```go
r.tp.Lock() // a new lock protecting this from concurrent modification
r.tracesProcessed = (r.tracesProcessed + 1) % 1000
r.tracesProcessed <= r.percentage {
  decision = Sampled
}
r.tp.Unlock()
```

This way, you don't need to keep track of the number of sampled traces. But do take concurrency into account: you'll need a write-lock around the assignment of r.tracesProcessed, reading the value back after that and assigning to a local v

jpkrohling commented on processor/tailsamplingprocessor/internal/sampling/percentage_test.go:
```diff
@@ -0,0 +1,103 @@
+		trace.Decisions = []Decision{NotSampled, Sampled}
+		decision, err = percentageFilter.Evaluate(traceID, trace)
+		assert.NoError(t, err)
+		assert.Equal(t, decision, NotSampled)
```
This test here is a bit confusing, running the same trace against the same filter twice. It would be easier to understand the expected behavior if you created a table-test with scenarios, like these:

1. 100 spans, 90 errors, 10% sampling => 90 + 1 traces
2. 100 spans, 10 errors, 10% sampling => 10 + 9 traces

Commit:
Use sampling_percentage similar to the probabilistic sampling processor

Use sampling_percentage instead of percentage
Make sampling_percentage a value between 0 and 100

Commit:
Combine tests in a single table-test

Commit:
Rename percentage -> probabilistic for consistency with probabilistic processor

Commit:
Add IncludeAlreadySampled option

Commit:
Clarify test cases

Commit:
Use the same algorithm as the probabilistic sampling processor

kvrhdn commented:
Looking at how the probabilistic sampling processor is implemented, I went for a very similar approach: instead of keeping a counter we can sample based upon the hashed trace ID. This is stateless (easier code) and allows the tail sampling processor to run distributed (i.e. if you have multiple collectors behind the [Trace ID aware load-balancing exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/loadbalancingexporter)).

This does not take into account already sampled traces, so the effective sampling percentage might be higher if other policies sample a lot. But the processor has the same implementation using `sampling.priority`.

Example:

```
policies:
  latency:        samples 10% of traces
  probabilistic:  samples 25% of traces

effective sample percentage: 10% + (25% of remaining 90%) = 32,5%
```

I dislike that I basically had to copy a large part of the probabilistic sampling processor. Maybe we can extract the hashing and sampling part into a shared library? otel-go also does trace ID-based sampling and might also profit from this.

Commit:
Simplify if-else case

jpkrohling commented on processor/tailsamplingprocessor/README.md:
```diff
@@ -73,3 +79,21 @@ processors:
+Running the probabilistic sampling processor is more efficient than the tail sampling processor.
+The probabilistic sampling policy makes decision based upon the trace ID, so waiting until more spans have arrived will not influence its decision. 
+
+...you are already using the tail sampling processor: add the probabilistic sampling policy as last in your chain.
```
Is adding the probabilistic sampling policy at the head of the chain considered a configuration error? If so, should a warning be generated?

kvrhdn replied:
Hmm, after thinking some more this isn't actually true: the order does not matter because the tail sampling processor will always run all policies. If one of the policies returns `Sampled` it will keep the trace. I was thinking too much of this as a pipeline.

jpkrohling commented:
>instead of keeping a counter we can sample based upon the hashed trace ID

Jaeger does something very similar under the name of downsampling at the storage plugin: https://github.com/jaegertracing/jaeger/blob/c5642b708be9e2577cc1c494889ae97946ccc78a/storage/spanstore/downsampling_writer.go#L131-L140

It uses fnv64a, which would be my preference for this one here as well.

>Maybe we can extract the hashing and sampling part into a shared library? otel-go also does trace ID-based sampling and might also profit from this.

Great idea. I think we are using CRC for the load balancing exporter, but we should settle for the same across the board. My choice would be fnv instead of Murmur, though.

Commit:
Typo

Commit:
Order of tail sampling policies does not matter

kvrhdn commented:
> It uses fnv64a, which would be my preference for this one here as well.

Sounds good :+1: The implementation will be easier as well since it's in the standard library.

> Would you prefer to work on the refactoring first, and then change this PR, or to get this PR merged first, and do the refactoring later?

How about I switch this sampler to fnv64a, then we can (finally) merge this PR and look at refactoring/combining logic?

jpkrohling commented:
Sounds goos to me.

Commit:
Switch hashing implementation from murmur3 to fnv-1a

kvrhdn commented:
Not sure why CI is failing, might be a performance issue.

Tests work locally but take 13s, I'm guessing the runners are slower than my machine.
```
➜ go test -race ./...
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor	1.911s
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/idbatcher	0.848s
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling	13.687s
```

I'll lower the size of the test.

Commit:
Lower amount of traces to sample in test to speed up tests

bogdandrutu commented:
Please rebase

Commit:
Merge branch 'main' into percentage_sampling

# Conflicts:
#	CHANGELOG.md

kvrhdn commented:
Done :+1: Would like me to squash the commits or does the repo squash on merge?
